### DISCUSSION TITLE
Race Condition?

### DISCUSSION THREAD

**Discussion by dlgreenwald** [netalertx/NetAlertX#303] [General]
When editing a device name, the change doesn't seem to persist...maybe even most of the time.  The change will say it's successful, jumping back to the devices view will show it as successful, but then a few minutes later I'll refresh and notice that it's changed back to whatever the name was originally detected to be.  Every once in a while it'll take...and once that happens it sticks.  

I didn't create an Issue for this yet, I wanted to make sure this isn't expected behavior where Pholus always overwrites the name with what it detects or something?  But again it sticks sometimes...so maybe not.

**Comment by jokob-sk** [comment]
Hey!

This was reported previously but I couldn't reproduce it:

https://github.com/jokob-sk/Pi.Alert/issues/259

Happy for you to try to investigate?

Thanks in advance!
j

**Reply by dlgreenwald** [reply]
I'm not quite sure how to start investigating it.  I too tried disabling the scans and making changes and that seemed to help, but I had significant issues getting the scans to actually stop.  

Watching the logs scroll...is Pholus somehow blocking or being run in a synchronous way?  This is going to be a bit awkward to describe but when I started tailing the logs I saw the following:

```
21:40:16 [Process Scan]  Processing scan results
21:40:16 [Process Scan] Print Stats
21:40:16 [Scan Stats]    Devices Detected.......: 0
21:40:16 [Scan Stats]        arp-scan detected..: 0
21:40:16 [Scan Stats]        Pi-hole detected...: +0
21:40:16 [Scan Stats]        New Devices........: 0
21:40:16 [Scan Stats]    Devices in this cycle..: 0
21:40:16 [Scan Stats]        Down Alerts........: 0
21:40:16 [Scan Stats]        New Down Alerts....: 0
21:40:16 [Scan Stats]        New Connections....: 0
21:40:16 [Scan Stats]        Disconnections.....: 0
21:40:16 [Scan Stats]        IP Changes.........: 0
21:40:16 [Process Scan] Stats end
21:40:16 [Process Scan] Updating DB Info
```
It stayed there with no addition log messages written for a minute or so and spit out the following:
```
21:40:16 [Process Scan] Sessions Events (connect / discconnect)
21:40:16 [Process Scan] Creating new devices
21:40:16 [Process Scan] Updating Devices Info
21:40:16 [Process Scan] Resolve devices names
21:40:16 [PholusScan] Scan: Pholus for 120s (2.0min)
21:40:16 [PholusScan] Pholus scan on [interface] eth0 [mask] 192.168.68.0/24
21:40:57 [PholusScan] Scan: Pholus SUCCESS
21:41:33 [Update Device Name] Trying to resolve devices without name
21:41:33 [Update Device Name] Pholus entries from prev scans: 4972
21:42:42 [Update Device Name] Names Found (DiG/Pholus): 0 (0/0)
21:42:42 [Update Device Name] Names Not Found         : 17
```
You can see that as far as the logging is concerned they all printed at the same second.  It appears that once Pholus finished it printed the log message, and completed.

Watching it for a bit longer tailing the docker logs appears to consistently happen in bursts, this may be a false lead.

Some other odd race like behavior I occasionally see.  
* When fetching the device details, the UI will load stale data.  I'll hit refresh and it'll get the right data.  I caught this one too, the only difference between the two screens is hitting refresh:
![Capture3](https://github.com/jokob-sk/Pi.Alert/assets/6752538/3fd48bf0-e0b7-460d-b4f9-959a3450db86)
![Capture4](https://github.com/jokob-sk/Pi.Alert/assets/6752538/669258f0-6061-4e4b-956d-d39b1eabff4f)


* Despite having a specified sort order, hitting refresh on the All devices page changes the sort order.  Here is a screenshot:
![Capture](https://github.com/jokob-sk/Pi.Alert/assets/6752538/6415710d-e5ec-4403-be6f-bad276c5ed67)
![Capture1](https://github.com/jokob-sk/Pi.Alert/assets/6752538/e19dce42-3e00-4f4a-b70a-3486b0e6ecc2)

**Reply by dlgreenwald** [reply]
I tried moving the database to an SSD and am still running into this issue.  Honestly, and this is intuition speaking, I'm leaning to the problem being the multiple readers/writers to the sqllite database.  I know their documentation says that it can handle multiple writers by locking the entire database, but having been attempting to make changes extensively it seems like one of the python processes reads the database, then writes over it with it's state, rather than reading it clean.  Like the lock happens in the wrong place or my filesystem/os isn't providing filesystem features for managing writes (https://www.sqlite.org/faq.html#q5).  

Are you opposed to migrating to PostgressSQL or MySQL as the database in the docker compose?  It might handle the backend and frontend simultaneous changes better.  It looks like you have most of the database access wrapped so the change shouldn't be too onerous.  I think there are two python files which would need modified, and one php file along with the compose file and the dockerfile.

**Comment by jokob-sk** [comment]
Migrating to a different DB is not something I'm considering as I'm unsure if the issue gets solved that way. Also, it's not that big of an issue and this can be mitigated by running Pholus less frequently.

I like the simplicity of a SQLlite deployment as well as backups are more straightforward. 

If rather solve the underlying issue or build in a delay for the Pholus scan.

**Reply by dlgreenwald** [reply]
Fair enough.

As I mentioned I'm struggling to turn off the scanning, and I still see the overwrite occur so I'm not sure that it mitigates by running Pholus less frequently (unless there is a seperate bug that is causing Pholus to run more or less continuously).  I believe I have every possible scan and/or plugin disabled and the instance is still pretty much continuously scanning.  My current config is at the bottom.  The following is some selected lines of logs showing a couple things.  1) the near continuous scanning with no significant delay between entering wait, and starting.  2) each trip through is spending ~60 seconds doing something in between `Pholus entries from prev scans` and `Names Found`

Questions:
Is that main loop meant to be continuous or is it expected to sleep for some period as per the config (?Scan cycle delay?)?
Why is Pholus running/What is Pholus doing for 60 seconds if it's not running every iteration of the loop?

I'm willing to look into those questions myself, but a little pointer as to where in the code I'm looking for the main loop?  How is scheduling done? 

```
13:50:45 [updateState] changing state to: "Process: Wait"
13:50:50 [updateState] changing state to: "Process: Start"
13:50:50 [Update Device Name] Pholus entries from prev scans: 747
13:50:50 [Update Device Name] Trying to resolve devices without name
13:50:50 [Process Scan] Resolve devices names
13:51:59 [Update Device Name] Names Found (DiG/Pholus): 0 (0/0)
13:51:59 [Update Device Name] Names Not Found         : 17
13:51:59 [updateState] changing state to: "Process: Wait"
13:51:59 [MAIN] Process: Wait
13:52:04 [MAIN] Starting loop
13:52:04 [updateState] changing state to: "Process: Start"
13:52:04 [Process Scan] Resolve devices names
13:52:04 [Update Device Name] Trying to resolve devices without name
13:52:04 [Update Device Name] Pholus entries from prev scans: 747
13:53:13 [Update Device Name] Names Found (DiG/Pholus): 0 (0/0)
13:53:13 [Update Device Name] Names Not Found         : 17
13:53:13 [updateState] changing state to: "Process: Wait"
13:53:13 [MAIN] Process: Wait
13:53:18 [MAIN] Starting loop
13:53:18 [updateState] changing state to: "Process: Start"
13:53:18 [Process Scan] Resolve devices names
13:53:18 [Update Device Name] Trying to resolve devices without name
13:53:18 [Update Device Name] Pholus entries from prev scans: 747
13:54:27 [Update Device Name] Names Found (DiG/Pholus): 0 (0/0)
13:54:27 [Update Device Name] Names Not Found         : 17
13:54:27 [updateState] changing state to: "Process: Wait"
13:54:27 [MAIN] Process: Wait
13:54:32 [MAIN] Starting loop
13:54:32 [updateState] changing state to: "Process: Start"
13:54:32 [Update Device Name] Pholus entries from prev scans: 747
13:54:32 [Update Device Name] Trying to resolve devices without name
13:54:32 [Process Scan] Resolve devices names
13:55:41 [Update Device Name] Names Found (DiG/Pholus): 0 (0/0)
13:55:41 [Update Device Name] Names Not Found         : 17
```

##Current Config
```
771087f4c011:/home/pi/pialert/config# cat pialert.conf 
#-----------------AUTOGENERATED FILE-----------------#
#                                                    #
#         Generated:  2023-08-03_22-51-26            #
#                                                    #
#   Config file for the LAN intruder detection app:  #
#      https://github.com/jokob-sk/Pi.Alert          #
#                                                    #
#-----------------AUTOGENERATED FILE-----------------#


# General
#---------------------------
ENABLE_ARPSCAN=False
SCAN_SUBNETS=['192.168.68.0/24 --interface=eth0']
LOG_LEVEL='debug'
TIMEZONE='America/New_York'
ENABLE_PLUGINS=False
PLUGINS_KEEP_HIST=10000
PIALERT_WEB_PROTECTION=False
PIALERT_WEB_PASSWORD='8d969eef6ecad3c29a3a629280e686cf0c3f5d5a86aff3ca12020c923adc6c92'
INCLUDED_SECTIONS=['internet','new_devices','down_devices','events']
SCAN_CYCLE_MINUTES=15
REPORT_DASHBOARD_URL='https://pialert.internal.dlgreen.com'
DIG_GET_IP_ARG='-4 myip.opendns.com @resolver1.opendns.com'
UI_LANG='English'
UI_PRESENCE=['online','offline','archived']
DAYS_TO_KEEP_EVENTS=90
HRS_TO_KEEP_NEWDEV=0


# Email
#---------------------------
REPORT_MAIL=False
SMTP_SERVER='smtp.gmail.com'
SMTP_PORT=587
REPORT_TO='user@gmail.com'
REPORT_FROM='Pi.Alert <user@gmail.com>'
SMTP_SKIP_LOGIN=False
SMTP_USER='user@gmail.com'
SMTP_PASS='password'
SMTP_SKIP_TLS=False
SMTP_FORCE_SSL=False


# Webhooks
#---------------------------
REPORT_WEBHOOK=False
WEBHOOK_URL='http://n8n.local:5555/webhook-test/aaaaaaaa-aaaa-aaaa-aaaaa-aaaaaaaaaaaa'
WEBHOOK_PAYLOAD='json'
WEBHOOK_REQUEST_METHOD='GET'
WEBHOOK_SIZE=1024


# Apprise
#---------------------------
REPORT_APPRISE=False
APPRISE_HOST='http://localhost:8000/notify'
APPRISE_URL='mailto://smtp-relay.sendinblue.com:587?from=user@gmail.com&name=apprise&user=user@gmail.com&pass=password&to=user@gmail.com'
APPRISE_PAYLOAD='html'


# NTFY
#---------------------------
REPORT_NTFY=False
NTFY_HOST='https://ntfy.sh'
NTFY_TOPIC='replace_my_secure_topicname_91h889f28'
NTFY_USER='user'
NTFY_PASSWORD='passw0rd'


# PUSHSAFER
#---------------------------
REPORT_PUSHSAFER=False
PUSHSAFER_TOKEN='ApiKey'


# MQTT
#---------------------------
REPORT_MQTT=False
MQTT_BROKER='192.168.1.2'
MQTT_PORT=1883
MQTT_USER='mqtt'
MQTT_PASSWORD='passw0rd'
MQTT_QOS=0
MQTT_DELAY_SEC=2


# DynDNS
#---------------------------
DDNS_ACTIVE=False
DDNS_DOMAIN='your_domain.freeddns.org'
DDNS_USER='dynu_user'
DDNS_PASSWORD='A0000000B0000000C0000000D0000000'
DDNS_UPDATE_URL='https://api.dynu.com/nic/update?'


# PiHole
#---------------------------
PIHOLE_ACTIVE=False
DHCP_ACTIVE=False


# Pholus
#---------------------------
PHOLUS_ACTIVE=False
PHOLUS_TIMEOUT=120
PHOLUS_FORCE=False
PHOLUS_RUN='disabled'
PHOLUS_RUN_TIMEOUT=600
PHOLUS_RUN_SCHD='0 4 * * *'
PHOLUS_DAYS_DATA=7


# Nmap
#---------------------------
NMAP_ACTIVE=False
NMAP_TIMEOUT=150
NMAP_RUN='disabled'
NMAP_RUN_SCHD='0 2 * * *'
NMAP_ARGS='-p -10000'


# API
#---------------------------
API_CUSTOM_SQL='SELECT * FROM Devices WHERE dev_PresentLastScan = 0'


# DHCPLSS
#---------------------------
DHCPLSS_RUN='disabled'
DHCPLSS_CMD='python3 /home/pi/pialert/front/plugins/dhcp_leases/script.py paths={paths}'
DHCPLSS_paths_to_check=['/mnt/dhcp1.leases','/mnt/dhcp2.leases']
DHCPLSS_RUN_SCHD='0 2 * * *'
DHCPLSS_RUN_TIMEOUT=5
DHCPLSS_WATCH=['Watched_Value1','Watched_Value4']
DHCPLSS_REPORT_ON=['new','watched-changed']


# DHCPSRVS
#---------------------------
DHCPSRVS_RUN='disabled'
DHCPSRVS_CMD='python3 /home/pi/pialert/front/plugins/dhcp_servers/script.py'
DHCPSRVS_RUN_SCHD='0 2 * * *'
DHCPSRVS_RUN_TIMEOUT=5
DHCPSRVS_WATCH=['Watched_Value1']
DHCPSRVS_REPORT_ON=['new','watched-changed']


# NEWDEV
#---------------------------
NEWDEV_dev_MAC=''
NEWDEV_dev_Name='(unknown)'
NEWDEV_dev_Vendor=''
NEWDEV_dev_FirstConnection=''
NEWDEV_dev_LastConnection=''
NEWDEV_dev_LastIP=''
NEWDEV_dev_SkipRepeated=0
NEWDEV_dev_LastNotification=''
NEWDEV_dev_Network_Node_port='0'


# NMAPSRV
#---------------------------
NMAPSRV_RUN='disabled'
NMAPSRV_CMD='SELECT  ns.MAC as Object_PrimaryID, cast({s-quote}http://{s-quote} || dv.dev_LastIP as VARCHAR(100)) || {s-quote}:{s-quote} || cast( SUBSTR(ns.Port ,0, INSTR(ns.Port , {s-quote}/{s-quote})) as VARCHAR(100)) as Object_SecondaryID,  datetime() as DateTime,  ns.Service as Watched_Value1, ns.State as Watched_Value2, dv.dev_Name as Watched_Value3,        {s-quote}null{s-quote} as Watched_Value4,        ns.Extra as Extra, ns.MAC as ForeignKey FROM (SELECT * FROM Nmap_Scan) ns left JOIN (SELECT dev_Name, dev_MAC, dev_LastIP FROM Devices) dv  ON ns.MAC = dv.dev_MAC'
NMAPSRV_RUN_SCHD='0 2 * * *'
NMAPSRV_WATCH=['Watched_Value1']
NMAPSRV_REPORT_ON=['new','watched-changed']


# SNMPDSC
#---------------------------
SNMPDSC_RUN='disabled'
SNMPDSC_CMD='python3 /home/pi/pialert/front/plugins/snmp_discovery/script.py routers={s-quote}{routers}{s-quote}'
SNMPDSC_routers=['snmpwalk -v 2c -c public -OXsq 192.168.1.1 .1.3.6.1.2.1.3.1.1.2']
SNMPDSC_RUN_SCHD='0 2 * * *'
SNMPDSC_RUN_TIMEOUT=5
SNMPDSC_WATCH=['Watched_Value1']
SNMPDSC_REPORT_ON=['new','watched-changed']


# UNDIS
#---------------------------
UNDIS_RUN='disabled'
UNDIS_CMD='python3 /home/pi/pialert/front/plugins/undiscoverables/script.py devices={devices}'
UNDIS_RUN_TIMEOUT=10
UNDIS_devices_to_import=['dummy_router']


# UNFIMP
#---------------------------
UNFIMP_RUN='disabled'
UNFIMP_CMD='python3 /home/pi/pialert/front/plugins/unifi_import/script.py username={username} password={password}  host={host} sites={sites}  protocol={protocol} port={port} version={version}'
UNFIMP_username=''
UNFIMP_password=''
UNFIMP_protocol='https://'
UNFIMP_host='192.168.1.1'
UNFIMP_port='8443'
UNFIMP_version=''
UNFIMP_sites=['default']
UNFIMP_RUN_SCHD='0 2 * * *'
UNFIMP_RUN_TIMEOUT=5
UNFIMP_WATCH=['Watched_Value1','Watched_Value4']
UNFIMP_REPORT_ON=['new','watched-changed']


# WEBMON
#---------------------------
WEBMON_RUN='disabled'
WEBMON_CMD='python3 /home/pi/pialert/front/plugins/website_monitor/script.py urls={urls}'
WEBMON_RUN_SCHD='0 2 * * *'
WEBMON_API_SQL='SELECT * FROM plugin_website_monitor'
WEBMON_RUN_TIMEOUT=5
WEBMON_WATCH=['Watched_Value1']
WEBMON_REPORT_ON=['new','watched-changed']
WEBMON_urls_to_check=['https://google.com','https://duck.com']
WEBMON_SQL_internet_ip='SELECT dev_LastIP FROM Devices WHERE dev_MAC = {s-quote}Internet{s-quote}'


#-------------------IMPORTANT INFO-------------------#
#   This file is ingested by a python script, so if  #
#        modified it needs to use python syntax      #
#-------------------IMPORTANT INFO-------------------#
```

**Reply by jokob-sk** [reply]
Thanks so much for the help! I'd love to dedicate more time to the project, but currently, it's about ~1h a day on average, which is limiting. 

## Pholus config

Pholus is only currently used for name resolution, not device discovery 

According to your config, Pholus should now be disabled:

```
# Pholus
#---------------------------
PHOLUS_ACTIVE=False
PHOLUS_TIMEOUT=120
PHOLUS_FORCE=False
PHOLUS_RUN='disabled'
PHOLUS_RUN_TIMEOUT=600
PHOLUS_RUN_SCHD='0 4 * * *'
PHOLUS_DAYS_DATA=7
```

I haven't noticed any race condition behavior recently and my setup runs Pholus once a day or once a new device is found. 

e.g. for Schedule:

```
# Pholus
#---------------------------
PHOLUS_ACTIVE=False
PHOLUS_TIMEOUT=120
PHOLUS_FORCE=False
PHOLUS_RUN='schedule' # here
PHOLUS_RUN_TIMEOUT=600
PHOLUS_RUN_SCHD='0 4 * * *'
PHOLUS_DAYS_DATA=7
```

## SCAN_CYCLE_MINUTES=15

This setting is ONLY related to arp-scan, pihole, DHCP_ACTIVE scans.  (I added it into the description)

## Development vision 🔮

I'm slowly planning to transform all scans into Plugins to make the code more maintainable and readable. That would also mean that these scans should be converted in the future (implemented before the [Plugins system](https://github.com/jokob-sk/Pi.Alert/tree/main/front/plugins) was a thing):

Arp-Scan, Pholus, DHCP_Leases, NMAP

For that, I'm updating Plugin and Settings capabilities so it's easier for me to switch those over. I'm also trying to keep setups, settings, backward compatible. 

The __main__.py loop has a non-configurable delay of 5s:

```python
time.sleep(5) # wait for N seconds  
```

Some more context-providing docs on the app structure:

https://github.com/jokob-sk/Pi.Alert/blob/main/docs/DATABASE.md
https://github.com/jokob-sk/Pi.Alert/blob/main/pialert/README.md
 
Any help would be greatly appreciated,
Thanks,
j

**Comment by jokob-sk** [comment]
I'm also moving a lot of the DB access to the API endpoints so the DB isn't used as frequently as it is right now.

**Reply by dlgreenwald** [reply]
I think this is likely a great call, as at a minimum it'll reduce the number of clients trying to read from the database with their own file handles.

**Comment by dlgreenwald** [comment]
I think I found the race.  In device.py update_devices_names.  The first thing it does is select all the `(unknown)` or `(name not found)` records.

It then does stuff that can take quite a bit of time, in my case its taking 60 seconds or so.

And then *without checking to see if the record has been updated in the meantime* writes over the top of the records with names that were not found (line 464) with the name `(Name Not Found)`

Combining this with the 5 second sleep in the main loop means that unless you write in that 5 seconds, any change is going to be overwritten.  Situation is worse the more `(Unknown)` or `(name not found)` records you have likely because of the dig call.

I can work on a patch tonight and submit a pull request.  I'm thinking of an approach where rather than validate each record hasn't changed instead only update the records where the name is changing.  So `(unknown)` to `(name not found)`.  This would mean that an edit of a `(unknown)` record would still get overwritten but those are short lived.

**Reply by jokob-sk** [reply]
Amazing ❤ Thanks a lot - apologies you had to figure the app loop out yourself - it was my bedtime 😴
PR approved

**Comment by dlgreenwald** [comment]
Followup to my own question:
`Is that main loop meant to be continuous or is it expected to sleep for some period as per the config (?Scan cycle delay?)?`
I tried setting this to conf.SCAN_CYCLE_DELAY and it almost works as expected.  The one hitch is that the preferences don't get updated in the database until that loop runs so the preference page keeps showing the wrong thing and warning you about it.  I think it might be work rethinking that aspect of the main loop so that when idle it's really idle, rather than waking up every few seconds to load the config only to go back to sleep.

**Reply by jokob-sk** [reply]
Do you think it would cause too frequent DB access? That's the main reason I incorporated the 5s delay.

**Reply by dlgreenwald** [reply]
Since the DB access would would only happen when the user viewed or saved the preferences (and it's already doing the read) I think that might actually be less DB access than waking up and reading the status every 5 seconds to determine if config has updated.

**Reply by jokob-sk** [reply]
Alright, I'll give it a go and test it

**Reply by jokob-sk** [reply]
Alright, I'd have to rewrite the main loop as there is a lot of writes to files (E.g. plugins, logging, etc) which would be killing HDD performance. So for now reverting back to the 5s delay.

**Comment by tismofied** [comment]
Thank you both very much for your hard work on this. I noticed this as well and I thought it was normal that names would revert back to default after scans.
